File: lessons/213/go-app/devices.go
package main

import (
	"context"
	"time"

	"github.com/prometheus/client_golang/prometheus"
	"go.mongodb.org/mongo-driver/mongo"
)

// Device represents hardware device
type Device struct {

	// Universally unique identifier
	UUID string `json:"uuid"`

	// Mac address
	Mac string `json:"mac"`

	// Firmware version
	Firmware string `json:"firmware"`
}

// Save inserts a Device into the MongoDb database.
func (d *Device) Save(ctx context.Context, db *mongo.Database, m *metrics) (err error) {
	// Get the current time to record the duration of the request.
	now := time.Now()
	defer func() {
		if err == nil {
			// Record the duration of the insert query.
			m.duration.With(prometheus.Labels{"op": "db"}).Observe(time.Since(now).Seconds())
		}
	}()

	_, err = db.Collection("go_devices").InsertOne(ctx, d)
	return annotate(err, "db.Exec failed")
}

File: lessons/239-fastapi-vs-go-golang-vs-nodejs-performance-price/node-app/devices.js
import sql from "./db.js";

// Inserts a Device into the Postgres database.
async function save({ uuid, mac, firmware, createdAt, updatedAt }) {
  return sql`INSERT INTO node_device (uuid, mac, firmware, created_at, updated_at) VALUES (${uuid}, ${mac}, ${firmware}, ${createdAt}, ${updatedAt}) RETURNING id;`;
}

export default save;
